lit_gpt/lora.py

```diff
@@ -151,36 +151,17 @@ def reset_parameters(self):
             nn.init.kaiming_uniform_(self.lora_A, a=math.sqrt(5))
             nn.init.zeros_(self.lora_B)
 
-    def train(self, mode: bool = True):
-        """Set the module into train or eval mode.
-
-        Args:
-            mode: if True the module will be set into train mode, if False - eval mode.
-        """
+    def merge(self):
+        """Merges the LoRA weights into the full-rank weights (W = W + delta_W)."""
 
         def T(w):
             return w.transpose(0, 1) if self.fan_in_fan_out else w
 
-        # despite being called from nn.Linear this method will put all layers into train mode, including nn.Dropout
-        # of course except parameters (such as self.lora_A, self.lora_B)
-        super().train(mode)
-
-        # if we want to put the layer into `train` mode then subtract LoRA weights if weights are already merged, so we
-        # can keep original weights untouched and train LoRA's matrices A and B separately.
-        # if we want to put into 'eval` mode - merge pretrained weights with LoRA's matrices A and B (if it's not
-        # already done) to reduce computation overhead during inference.
-        if mode:
-            if self.merge_weights and self.merged:
-                # Make sure that the weights are not merged
-                if self.r > 0:
-                    self.weight.data -= T(self.lora_B @ self.lora_A) * self.scaling
-                self.merged = False
-        else:
-            if self.merge_weights and not self.merged:
-                # Merge the weights and mark it
-                if self.r > 0:
-                    self.weight.data += T(self.lora_B @ self.lora_A) * self.scaling
-                self.merged = True
+        if self.merge_weights and not self.merged:
```

Comment:
I feel somewhat uncomfortable about it. 
If someone explicitly calls `merge` method - that doesn't necessarily mean that the weights will be indeed merged:
1. if `self.merge_weights` is False 
2. if `self.r` <= 0

... and it will happen without any warnings.

But also the question is do we need those checks?
1. `self.merge_weights` perhaps was a way to prevent automatically merging weights in `train` method. Since now merging is done explicitly we don't need it.
2. `self.r = -1` is a way to disable LoRA weights and get a regular Linear layer, but we don't use this functionality. For me it look like using `nn.Linear` if we don't need LoRA functionality is simpler (more obvious). But I might be wrong.
So we could instead of:
```python
if config.to_projection:
    self.proj = LoRALinear(
        config.n_embd,
        config.n_embd,
        bias=config.bias,
        r=config.r,
        lora_alpha=config.alpha,
        lora_dropout=config.dropout,
    )
else:
    self.proj = nn.Linear(config.n_embd, config.n_embd, bias=config.bias)
```
use this
```python
self.proj = LoRALinear(
    config.n_embd,
    config.n_embd,
    bias=config.bias,
    r=config.r if config.to_projection else -1,
    lora_alpha=config.alpha,
    lora_dropout=config.dropout,
)
```
---
So I think either at least add some warning information or get rid of `merge_weights` and assert that `self.r` is greater than 0.

What do you think @awaelchli ?
I tend to over-complicate things, so I'm relying on your input.

Reply:
I am not concerned about it, we use `merge_weights=True` everywhere so I don't think this can have negative impacts. The code for this was inherited from the original lora implementation: https://github.com/microsoft/LoRA/blob/998cfe4d351f4d6b4a47f0921dec2397aa0b9dfe/loralib/layers.py#L65
If r = 0, that means LoRA is disabled (delta_w = 0) so merging being a no-op is not unreasonable. Whether this should error/be allowed or not is merely a design decision IMO. 
I'm not opposed to removing the merge_weights attribute because it's not useful anymore, and we don't use it in our scripts.

Reply:
I'll create a PR if you don't mind.